### ISSUE TITLE
[Feature Request] Routes implementation and notifications

### ISSUE THREAD

**Issue by inmsa2014** [tananaev/traccar#3761]
I have used other platforms, and once worked with routes which is a very useful option. This might be an interesting implementation for Traccar as well.

1. Draw a route (multiple origins and destinations for complex routes with many pickups and deliveries).
       ...This might work using one or more geofences for origins and the same for destinations.

2. Calculate a Polyline to be the route, and the possibility of changing the route manually, and we can input the radius of freedom to get out of the polilyne based on geoposition.

3.- Assign a route to an object, with a reference number, dates and times for arriving at origins and destinations, where once it is assigned, server will follow up determined object from begginning to end, and save in the database this route with geoposition and attributes for each point saved during the route tracking (kind of live tracking but for this new feature).

4.- ALERTS: Durign route tracking, we can set different notifications, like the following; Out of route alert, Calculate E.T.A. and Estimated distance of arrival, Fuel drop, using geofences with alerts as points prohibited for drivers (This can already be done with geofences), and measure using reports driver performance during determined routes, and it's all coming to my mind for now.

Please share your thoughts on implementation possibility on your code.

**Comment by tananaev** [comment]
https://www.traccar.org/forums/topic/feature-request-routes-implementation-and-notifications/

**Comment by Abyss777** [comment]
@inmsa2014 you have not share the purpose or use case of this suggestions.

I also check sometimes other platforms and see a lot of reports and other features. And I think most of them to run onetime and say "Wow, cool..." and never use it again.

As I can understand all your suggestions are about delivery service or merchandiser work. I do not think traccar have a course to implement full featured delivery service platform.

But If make traccar part of your delivery service platform, it already allow achieve most of suggestions with API.
Lets imagine, you already have platform to receive orders, make tasks to deliverer and others. It is possible to teach your platform generate geofences according to delivery plan, link them to appropriate devices/drivers, receive notifications about arriving to geofence and so on. You can also integrate this platform with some routing service (https://wiki.openstreetmap.org/wiki/Routing) to generate optimal routes and "Calculate E.T.A. and Estimated distance of arrival" 

All above is my IMHO

**Comment by inmsa2014** [comment]
@abyss777 you are right on what you say, I personally run a Logistics and Trucking company, what I have learned from Traccar is that we can offer our solutions at a relative low cost, taking advantage of open source software. I personally started to implement new ways of keeping our business running at a good pace, plus offering truck owner operators a completely free way to keep their assets tracked and supervised against theft down here in Mexico where we have a lot of issues regarding this matter. 

Regarding Traccar and it's course of development, I have read that it has gone better and better, obviously receiving and implementing useful features from hearing users and debating among developers. My post is a Feature Request for evaluation and comments, I am a new user and the great work I see it has been done, is something we value and will somehow support, either by posting about our opinions or any other ways available. I am not a programmer in any way, but I understand some about it, and it seems it may be a full time job for Anton and it's team, and we will contribute in any way we can.

About our request, I try to understand where Traccar is heading and we as a company, think it could be a Feature that may be able to help other people in our business and other Delivery business as well. For example we have human resources doing this job 24/7 it would be useful to automate this process using the asset tracking system that we use to send status about trucks directly from our servers to our customers, management, operation and customer departments. 

Thank you for your Comments. I hope ours are not seem badly in any way.

**Comment by kirantpatil** [comment]
Please find some of the earlier discussions on routing.

https://github.com/tananaev/traccar/issues/2629
https://github.com/tananaev/traccar/issues/2866

OSRM with travelling sales person for shortest route calculation.

https://github.com/Project-OSRM/osrm-backend/issues/1704
